File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def getRightMost(leftnode,curr):
    while(leftnode.right!=None and leftnode.right!=curr):
           leftnode=leftnode.right
    return leftnode
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        curr=root
        ans=[]
        while(curr!=None):
            leftnode=curr.left
            if(leftnode==None):
                ans.append(curr.val)
                curr=curr.right
            else:
                rightmost=getRightMost(leftnode,curr)
                if(rightmost.right==None):
                    rightmost.right=curr
                    curr=curr.left
                else:
                    rightmost.right=None
                    ans.append(curr.val)
                    curr=curr.right
        return ans

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode getRightMost(TreeNode leftnode,TreeNode curr){
        while (leftnode.right!=null && leftnode.right!=curr)
            leftnode=leftnode.right;
        return leftnode;
    }
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> arr=new ArrayList<Integer>();
        TreeNode curr=root;
        while(curr!=null){
            TreeNode leftnode=curr.left;
            if(leftnode==null){
                arr.add(curr.val);
                curr=curr.right;
            }
            else{
                TreeNode rightmost=getRightMost(leftnode,curr);
                if(rightmost.right==null){
                    rightmost.right=curr;
                    curr=curr.left;
                }
                else{
                    rightmost.right=null;
                    arr.add(curr.val);
                    curr=curr.right;
                }
            }
        }
        return arr;
    }
}